Полная информация об исключении при загрузке файлов с дымовыми тестами в случае конфигурации с поддержкой синхронности

BarinovIN opened the pull request:
частичная реализация  #1095

сделанные изменения

- Доработал обработку исключения при загрузке файла в загрузчике каталога. Метод "Сообщить" в итоге ничего не сообщал. Теперь причина пропуска файла кратко будет видна в логе выполнения и подробно в отладочном логе и ЖР.
- Вызываю исключение при ошибке в загрузке файла, т.к. уже нельзя будет считать, что все тесты выполнены успешно, даже если остальные файлы загрузятся и других ошибок не будет. Но не нашёл возможности отсюда взвести флаг ошибки для тестов в целом. Пропускать незагрузившийся файл без изменения статуса - искажение результатов тестирования.

@vanessa-opensource/Collaborators - просьба прокомментировать и проверить

BarinovIN committed:
Полная информация об исключении при загрузке файлов с дымовыми тестами

#1095

artbear commented on plugins/ЗагрузчикКаталога/ЗагрузчикКаталога/Forms/Форма/Ext/Form/Module.bsl:
```diff
@@ -189,7 +189,25 @@
+		ПолныйТекстОшибки = ПодробноеПредставлениеОшибки(ИнформацияОбОшибке);
+		КраткийТекстОшибки = КраткоеПредставлениеОшибки(ИнформацияОбОшибке);
+				
+		ПолныйТекстОшибки = КонтекстЯдра.СтрШаблон_(ШаблонСообщения, ФайлОбработки.ПолноеИмя, ПолныйТекстОшибки);
```
уже можно\нужно юзать простое СтрШаблон вместо КонтекстЯдра.СтрШаблон_

BarinovIN replied:
Заменил. Может в методе СтрШаблон_ тогда заменить тело на вызов простого СтрШаблон  и пометить метод как устаревший, чтобы было понятнее, что его уже использовать не нужно? Или я не до конца понимаю его назначение?

artbear commented on plugins/ЗагрузчикКаталога/ЗагрузчикКаталога/Forms/Форма/Ext/Form/Module.bsl:
```diff
@@ -189,7 +189,25 @@
+		КраткийТекстОшибки = КраткоеПредставлениеОшибки(ИнформацияОбОшибке);
+				
+		ПолныйТекстОшибки = КонтекстЯдра.СтрШаблон_(ШаблонСообщения, ФайлОбработки.ПолноеИмя, ПолныйТекстОшибки);
+		КраткийТекстОшибки = КонтекстЯдра.СтрШаблон_(ШаблонСообщения, ФайлОбработки.Имя, КраткийТекстОшибки);
```
аналогично заменить на СтрШаблон

artbear commented on plugins/ЗагрузчикКаталога/ЗагрузчикКаталога/Forms/Форма/Ext/Form/Module.bsl:
```diff
@@ -189,7 +189,25 @@
+		// Тут достаточно взвести флаг ошибки для всех тестов целиком и продолжить, но непонятно как взвести флаг ошибки отсюда.
+		// Поэтому пока вызываем исключение.
+		// Лучше ведь починить и перезапустить, чем пропустить ошибку из-за незагруженного файла с тестами.
+		ВызватьИсключение;
```
проверял в пакетном режиме?

имхо при вызове исключения здесь основной фреймворк тестирования может упасть,
а тогда сеанс 1С просто зависнет и останется в памяти, не завершившись.

а что происходит в интерактивном режиме?

BarinovIN replied:
#### В пакетном режиме.
Выводится сообщение:
```
Ошибка загрузки и выполнения тестов в пакетном режиме
И дальше подробная информация об ошибке
```
Менеджер тестирования завершается, а клиент действительно остаётся открыт. Но так он может оставаться и из-за других исключений.
Срабатывает вот этот [перехват исключения](https://github.com/vanessa-opensource/add/blob/develop/epf/xddTestRunner/xddTestRunner/Forms/УправляемаяФорма/Ext/Form/Module.bsl#L1415), возможно в нём стоит добавить закрытие клиентов тестирования так же, как [вот здесь](https://github.com/vanessa-opensource/add/blob/develop/epf/xddTestRunner/xddTestRunner/Forms/УправляемаяФорма/Ext/Form/Module.bsl#L1483). Тогда ничего зависшего оставаться не будет.

#### В интерактивном режиме.
Фреймворк не закрывается, появляется два сообщения.
- Короткое:
```
Не удалось загрузить файл ИмяФайла.epf
Причина исключения
```
- И подробное описание ошибки со стеком вызовов.

#### Итого
По-моему поведение приемлемое. Но как я уже писал, исключение в этой строке хотелось бы убрать, заменив на установку статуса ошибки для тестов в целом.

artbear reviewed (changes requested):
Важные вопросы

BarinovIN committed:
Уже нужно юзать простое СтрШаблон вместо КонтекстЯдра.СтрШаблон_

artbear commented on plugins/ЗагрузчикКаталога/ЗагрузчикКаталога/Forms/Форма/Ext/Form/Module.bsl:
```diff
@@ -189,7 +189,25 @@
 		
 	Исключение
-		Сообщить("Не удалось загрузить файл " + ФайлОбработки.ПолноеИмя + Символы.ПС + ПодробноеПредставлениеОшибки(ИнформацияОбОшибке()));
+		ШаблонСообщения = НСтр("ru = 'Не удалось загрузить файл %1
```
остается главный вопрос - почему исправление сделано в Загрузчике каталога, а не в загрузчике файла?

1 ведь основную работу выполняет именно загрузчик файла

2 и в коде обработчика из ПР используются те же данные, что передаются в загрузчик файла - полное имя файла

Предлагаю перенести код Попытки-Исключение в метод Загрузить загрузчика файлов

artbear replied:
@BarinovIN а в целом я с исправлением согласен, только нужно код перенести в загрузчик файла и тогда я смержу.

BarinovIN replied:
Исправление сделал там, где было необработанное исключение, добавил его полноценную обработку.
Другой вариант исправления - убрать перехват исключения. Он же только сообщение об ошибке уточняет - это не обязательно.
Необходим ли перехват исключения в загрузчике файла?

Итого, предлагаю совсем убрать перехват исключения отсюда, но и в загрузчик файла не добавлять его.

BarinovIN replied:
Перенёс перехват исключения в загрузчик файла.

artbear commented:
Уточнил заголовок, т.к. ПР решает проблему только для конфигураций с поддержкой синхронности

для асинхронных конфигураций не используются методы Загрузить, а используются методы вида НачатьЗагрузку.

а в них обработка сложнее, т.к. простого исключения недостаточно, остановится только текущая асинхронная функция, а не весь фреймворк ((

BarinovIN committed:
Перенести перехват исключения в загрузчик файла

Перехват исключения фактически нужен "на будущее".
Для возможности пропустить незагруженный файл и обработать остальные,
но со статусом ошибки для тестов в целом.
Пока установки такого статуса нету, поэтому будет вызвано исключение
и все тесты не выполнятся.
Для этих же целей дерево тестов нужно сформировать заранее,
чтобы функция всегда его возвращала, после замены вызова исключения
на установку статуса ошибки.

Сейчас перехват исключения поможет быстро узнать имя файла,
который не удалось загруить.

см: #1095
